stats/cloud/data.go

```diff
@@ -0,0 +1,134 @@
+				metrics.HTTPReqWaiting.Name:        stats.D(trail.Waiting),
+				metrics.HTTPReqReceiving.Name:      stats.D(trail.Receiving),
+			},
+		},
+	}
+}
+
+// SampleDataAggregatedHTTPReqs is used in aggregated samples for HTTP requests.
+type SampleDataAggregatedHTTPReqs struct {
+	Time   time.Time         `json:"time"`
+	Type   string            `json:"type"`
+	Count  uint64            `json:"count"`
+	Tags   *stats.SampleTags `json:"tags,omitempty"`
+	Values struct {
+		Duration       AggregatedMetric `json:"http_req_duration"`
+		Blocked        AggregatedMetric `json:"http_req_blocked"`
+		Connecting     AggregatedMetric `json:"http_req_connecting"`
+		TLSHandshaking AggregatedMetric `json:"http_req_tls_handshaking"`
+		Sending        AggregatedMetric `json:"http_req_sending"`
+		Waiting        AggregatedMetric `json:"http_req_waiting"`
+		Receiving      AggregatedMetric `json:"http_req_receiving"`
+	} `json:"values"`
+}
+
+// CalcAverages calculates and sets all `Avg` properties in the `Values` struct
+func (sdagg *SampleDataAggregatedHTTPReqs) CalcAverages() {
+	count := float64(sdagg.Count)
+	sdagg.Values.Duration.Avg = float64(sdagg.Values.Duration.sum) / count
+	sdagg.Values.Blocked.Avg = float64(sdagg.Values.Blocked.sum) / count
+	sdagg.Values.Connecting.Avg = float64(sdagg.Values.Connecting.sum) / count
+	sdagg.Values.TLSHandshaking.Avg = float64(sdagg.Values.TLSHandshaking.sum) / count
+	sdagg.Values.Sending.Avg = float64(sdagg.Values.Sending.sum) / count
+	sdagg.Values.Waiting.Avg = float64(sdagg.Values.Waiting.sum) / count
+	sdagg.Values.Receiving.Avg = float64(sdagg.Values.Receiving.sum) / count
+}
+
+// AggregatedMetric is used to store aggregated information for a
+// particular metric in an SampleDataAggregatedMap.
+type AggregatedMetric struct {
+	Min time.Duration `json:"min"`
+	Max time.Duration `json:"max"`
+	sum time.Duration `json:"-"`   // ignored in JSON output because of SampleDataAggregatedHTTPReqs.Count
+	Avg float64       `json:"avg"` // not updated automatically, has to be set externally
+}
+
+// Add the new duration to the internal sum and update Min and Max if necessary
+func (am *AggregatedMetric) Add(t time.Duration) {
+	if am.sum == 0 || am.Min > t {
+		am.Min = t
+	}
+	if am.Max < t {
+		am.Max = t
+	}
+	am.sum += t
+}
+
+type aggregationBucket map[*stats.SampleTags][]*netext.Trail
+
+type durations []time.Duration
+
+func (d durations) Len() int           { return len(d) }
+func (d durations) Swap(i, j int)      { d[i], d[j] = d[j], d[i] }
+func (d durations) Less(i, j int) bool { return d[i] < d[j] }
+func (d durations) GetNormalBounds(iqrCoef float64) (min, max time.Duration) {
+	l := len(d)
+	if l == 0 {
+		return
+	}
+
+	var q1, q3 time.Duration
+	if l%4 == 0 {
+		q1 = d[l/4]
+		q3 = d[(l/4)*3]
+	} else {
+		q1 = (d[l/4] + d[(l/4)+1]) / 2
+		q3 = (d[(l/4)*3] + d[(l/4)*3+1]) / 2
+	}
+
+	iqr := float64(q3 - q1)
+	min = q1 - time.Duration(iqrCoef*iqr)
+	max = q3 + time.Duration(iqrCoef*iqr)
+	return
+}
```

Comment:
:+1: for the sorting, and I'll fix it for `len(d) = 1` when I write some tests for it.

> I'd use linear calculation of q1 and q3.

You mean instead of taking the average of the two neighbouring values to linearly interpolate them in the `else` clause?

Reply:
Yep